osf/utils/machines.py

```diff
@@ -317,3 +346,170 @@ def get_context(self):
             'requester': self.machineable.creator,
             'is_request_email': True,
         }
+
+
+class RegistrationMachine(BaseMachine):
+    ActionClass = RegistrationRequestAction
+    States = RegistrationStates
+    Transitions = REGISTRATION_TRANSITIONS
+
+    def save_action(self, ev):
+        user = ev.kwargs.get('user')
+        self.action = self.ActionClass.objects.create(
+            target=self.machineable,
+            creator=user,
+            trigger=ev.event.name,
+            from_state=self.from_state.name,
+            to_state=ev.state.name,
+            comment=ev.kwargs.get('comment', ''),
+        )
+        self.machineable.save()
+
+    def resubmission_allowed(self, ev):
+        return self.machineable.provider.reviews_workflow == Workflows.PRE_MODERATION.value
```

Comment:
As discussed on Zoom: The logic of these callbacks should mostly be moved into functions on the DraftRegistration.

These functions are almost exclusively examining/modifying details of machineable, which suggests that the functionality belongs there to begin with. That makes it easier to discover/understand the guts of moderation process from the DraftRegistration itself, while the StateMachine is just responsible for describing requirements around transitions.

Going to keep these callback functions here to a) avoid fighting the existing pattern and b) keep awareness of the "event_data" concept out of the higher-level abstraction, but they will just call functions directly on machineable.

Reply:
I'm just going to delete this line and hopefully this logic will be re-implemented in the correct way in the next ticket.